### ISSUE TITLE
SGD based on Kalman Filter

### ISSUE THREAD

**Issue by Leo-VK** [online-ml/river#341]
Hi 

Recently, I came across a [Stochastic Gradient Descent version based on the Kalman Filter](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.418.7228&rep=rep1&type=pdf) (eq. 34-38) which could be interesting for creme. It uses the Hessian matrix and is said to converge considerably faster than regular Stochastic Gradient Descent. 
The aforementioned example describes the case for the least squares loss, but could be extended for any twice-differentiable cost function. Regarding practical implementation, the idea would be to implement a Kalman Filter optimizer in optim, such that various models could benefit.
A corse issue is the integration of the second derivative for the optimizer in __update_after_pred_. Would a function _eval_second_derivative_ for the corresponding losses be an option? However, the fit_one function would then need some changes which would add further complexity.

**Comment by MaxHalford** [comment]
Hey Leandro. 

The way to go about this would be to add a `hessian` method to each loss function in the [`losses.py`](https://github.com/creme-ml/creme/blob/master/creme/optim/losses.py) file. You could just start off by implementing it just for the `Squared` loss as a proof of concept. Once this is done, you could add a file at the root of the [`optim` directory](https://github.com/creme-ml/creme/blob/master/creme/optim/). Is this clear? :)

**Comment by Leo-VK** [comment]
Hi Max

This is pretty clear thanks. My question would rather concern the update formula:
Along your suggestions, the idea is to add a <code> kalman_filter</code> module in the <code> optim </code> directory and do the update calculations including hessian matrix assembly in the __update_after_pred_ method. Regarding the update, equation (38) is basically the shermann_morrison formula which you already provide in the <code>utils.math</code> module.
At least in the least squares case, we'd need access to features x from within the <code>kalman_filter</code> module in the <code>optim</code> direcotry. I'm not sure how to go about that without changing the __update_after_pred_ inputs.

**Comment by MaxHalford** [comment]
That's a good point. If `x` is really required, then we'll have to look into adding it to the signature of `update_after_pred`. Are you definitely sure that it's required?

To get started, I would suggest that you implement this yourself without worrying about how things are organised in `creme`. I can then take care of integrating it seamlessly. I think it's easier for everyone if you focus on getting good results, instead of getting bogged into API details :)
